cluster-autoscaler/cloudprovider/openstack/openstack_manager_heat.go

```diff
@@ -0,0 +1,304 @@
+	}
+
+	if cfg.Global.CAFile != "" {
+		roots, err := certutil.NewPool(cfg.Global.CAFile)
+		if err != nil {
+			return nil, err
+		}
+		config := &tls.Config{}
+		config.RootCAs = roots
+		provider.HTTPClient.Transport = netutil.SetOldTransportDefaults(&http.Transport{TLSClientConfig: config})
+
+	}
+
+	err = openstack.AuthenticateV3(provider, authOpts, gophercloud.EndpointOpts{})
+	if err != nil {
+		return nil, fmt.Errorf("could not authenticate: %v", err)
+	}
+
+	clusterClient, err := openstack.NewContainerInfraV1(provider, gophercloud.EndpointOpts{Type: "container-infra", Name: "magnum", Region: cfg.Global.Region})
+	if err != nil {
+		return nil, fmt.Errorf("could not create container-infra client: %v", err)
+	}
+
+	heatClient, err := openstack.NewOrchestrationV1(provider, gophercloud.EndpointOpts{Type: "orchestration", Name: "heat", Region: cfg.Global.Region})
+	if err != nil {
+		return nil, fmt.Errorf("could not create orchestration client: %v", err)
+	}
+
+	manager := OpenstackManagerHeat{
+		clusterClient: clusterClient,
+		clusterName:   opts.ClusterName,
+		heatClient:    heatClient,
+		waitTimeStep:  waitForStatusTimeStep,
+	}
+
+	manager.stackID, err = manager.GetStackID()
+	if err != nil {
+		return nil, fmt.Errorf("could not store stack ID on manager: %v", err)
+	}
+	manager.stackName, err = manager.GetStackName()
+	if err != nil {
+		return nil, fmt.Errorf("could not store stack name on manager: %v", err)
+	}
+
+	return &manager, nil
+}
+
+// NodeGroupSize gets the current cluster size as reported by magnum.
+// The nodegroup argument is ignored as this implementation of OpenstackManager
+// assumes that only a single node group exists.
+func (osm *OpenstackManagerHeat) NodeGroupSize(nodegroup string) (int, error) {
+	cluster, err := clusters.Get(osm.clusterClient, osm.clusterName).Extract()
+	if err != nil {
+		return 0, fmt.Errorf("could not get cluster: %v", err)
+	}
+	return cluster.NodeCount, nil
+}
+
+// UpdateNodeCount replaces the cluster node_count in magnum.
+func (osm *OpenstackManagerHeat) UpdateNodeCount(nodegroup string, nodes int) error {
+	updateOpts := []clusters.UpdateOptsBuilder{
+		UpdateOptsInt{Op: clusters.ReplaceOp, Path: "/node_count", Value: nodes},
+	}
+	_, err := clusters.Update(osm.clusterClient, osm.clusterName, updateOpts).Extract()
+	if err != nil {
+		return fmt.Errorf("could not update cluster: %v", err)
+	}
+	return nil
+}
+
+// GetNodes currently returns nothing and this seems to work fine.
+func (osm *OpenstackManagerHeat) GetNodes(nodegroup string) ([]string, error) {
+	// TODO: get nodes from heat?
+	// This works fine being empty for now anyway
+	return []string{}, nil
+}
+
+// DeleteNodes deletes nodes by passing a comma separated list of names or IPs
+// of minions to remove to heat, and simultaneously sets the new number of minions on the stack.
+// The magnum node_count is then set to the new value (does not cause any more nodes to be removed).
```

Comment:
Maybe we can add a reference here to the upstream storyboard:
https://storyboard.openstack.org/#!/story/2005052

and describe the usage of heat for this purpose is temporary until it gets implemented?

Reply:
+1
I think it would be nice to document the process/workflow how it works in OpenStack world. And how it interact with Magnum and Heat.

Reply:
Done [here](https://github.com/kubernetes/autoscaler/pull/1690/commits/005084be245456f7ac4a04204a8f284dc6b9bb9d).